File: frontend/kotlin/frontend-kotlin-compose/app/src/main/kotlin/dev/multistack/compose/data/Session.kt
package dev.multistack.compose.data

import android.content.Context
import android.content.SharedPreferences
import dev.multistack.compose.BuildConfig

/**
 * The SPA keeps the token in `localStorage` under `authToken:<backendId>`
 * (`lib/appBase.ts authTokenStorageKey`). SharedPreferences is the native
 * equivalent and reuses the same key, so a session is scoped per matrix backend
 * and never leaks across cells.
 */
class Session(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("multistack-auth", Context.MODE_PRIVATE)

    val tokenKey: String = BuildConfig.BACKEND_ID
        .takeIf { it.isNotBlank() }
        ?.let { "authToken:$it" }
        ?: "authToken"

    fun token(): String? = prefs.getString(tokenKey, null)?.takeIf { it.isNotBlank() }

    fun save(token: String) {
        prefs.edit().putString(tokenKey, token).apply()
    }

    fun clear() {
        prefs.edit().remove(tokenKey).apply()
    }
}

File: frontend/swift/frontend-swift-swiftui/Sources/Data/Session.swift
import Foundation

/// The SPA keeps the token in `localStorage` under `authToken:<backendId>`
/// (`lib/appBase.ts authTokenStorageKey`). `UserDefaults` is the native
/// equivalent and reuses the same key, so a session is scoped per matrix
/// backend and never leaks across cells.
struct Session {
    let tokenKey: String

    init(backendId: String = Config.backendId) {
        tokenKey = backendId.isEmpty ? "authToken" : "authToken:\(backendId)"
    }

    var token: String? {
        guard let value = UserDefaults.standard.string(forKey: tokenKey), !value.isEmpty else {
            return nil
        }
        return value
    }

    func save(_ token: String) {
        UserDefaults.standard.set(token, forKey: tokenKey)
    }

    func clear() {
        UserDefaults.standard.removeObject(forKey: tokenKey)
    }
}
